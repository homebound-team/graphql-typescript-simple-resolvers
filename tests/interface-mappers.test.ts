import { createTestSchema, runPlugin } from "./utils/test-helpers";

describe("Interface Mapped Types", () => {
  const schemaWithMappedInterfaces = `
    interface Publisher {
      name: String
      country: String
    }

    interface Identifiable {
      id: ID!
    }

    type SmallPublisher implements Publisher {
      name: String
      country: String
      employees: Int
    }

    type LargePublisher implements Publisher & Identifiable {
      id: ID!
      name: String
      country: String
      revenue: Float
      subsidiaries: [String!]!
    }

    type Book {
      title: String!
      publisher: Publisher
      isbn: String!
    }

    type Magazine {
      title: String!
      publisher: Publisher!
      issueNumber: Int!
    }

    type Query {
      publishers: [Publisher!]!
      identifiableItems: [Identifiable!]!
      getPublisher(id: ID!): Publisher
      books: [Book!]!
    }
  `;

  it("allows interfaces to be mapped to custom types", async () => {
    const schema = createTestSchema(schemaWithMappedInterfaces);
    const code = await runPlugin(schema, {
      mappers: {
        Publisher: "./entities#PublisherEntity",
        Identifiable: "./entities#IdentifiableEntity",
      },
      scalars: {},
      enumValues: {},
    });

    expect(code).toContain(`import { IdentifiableEntity, PublisherEntity } from "./entities";`);
  });

  it("generates mapped interface types", async () => {
    const schema = createTestSchema(schemaWithMappedInterfaces);
    const code = await runPlugin(schema, {
      mappers: {
        Publisher: "./entities#PublisherEntity",
      },
      scalars: {},
      enumValues: {},
    });

    // When interfaces are mapped, the plugin doesn't generate type aliases
    // but uses the mapped types directly in resolvers and references
  });

  it("generates types implementing mapped interfaces", async () => {
    const schema = createTestSchema(schemaWithMappedInterfaces);
    const code = await runPlugin(schema, {
      mappers: {
        Publisher: "./entities#PublisherEntity",
      },
      scalars: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface SmallPublisher {
  name: string | null | undefined;
  country: string | null | undefined;
  employees: number | null | undefined;
}`);

    expect(code).toContain(`export interface LargePublisher {
  id: string;
  name: string | null | undefined;
  country: string | null | undefined;
  revenue: number | null | undefined;
  subsidiaries: string[];
}`);
  });

  it("handles fields that reference mapped interfaces", async () => {
    const schema = createTestSchema(schemaWithMappedInterfaces);
    const code = await runPlugin(schema, {
      mappers: {
        Publisher: "./entities#PublisherEntity",
      },
      scalars: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface Book {
  title: string;
  publisher: PublisherEntity | null | undefined;
  isbn: string;
}`);

    expect(code).toContain(`export interface Magazine {
  title: string;
  publisher: PublisherEntity;
  issueNumber: number;
}`);
  });

  it("generates resolvers for types with mapped interface fields", async () => {
    const schema = createTestSchema(schemaWithMappedInterfaces);
    const code = await runPlugin(schema, {
      mappers: {
        Publisher: "./entities#PublisherEntity",
      },
      scalars: {},
      enumValues: {},
    });

    expect(code).toContain(`publisher: Resolver<Book, {}, PublisherEntity | null | undefined>;`);
  });

  it("generates interface resolvers for mapped interfaces", async () => {
    const schema = createTestSchema(schemaWithMappedInterfaces);
    const code = await runPlugin(schema, {
      mappers: {
        Publisher: "./entities#PublisherEntity",
      },
      scalars: {},
      enumValues: {},
    });

    expect(code).toContain(`Publisher: { __resolveType(o: SmallPublisher | LargePublisher): string };`);
  });

  it("generates query resolvers returning mapped interfaces", async () => {
    const schema = createTestSchema(schemaWithMappedInterfaces);
    const code = await runPlugin(schema, {
      mappers: {
        Publisher: "./entities#PublisherEntity",
      },
      scalars: {},
      enumValues: {},
    });

    expect(code).toContain(`publishers: Resolver<{}, {}, ReadonlyArray<PublisherEntity>>;`);
    expect(code).toContain(`getPublisher: Resolver<{}, QueryGetPublisherArgs, PublisherEntity | null | undefined>;`);
  });

  it("snapshots interface mapper generation", async () => {
    const schema = createTestSchema(schemaWithMappedInterfaces);
    const code = await runPlugin(schema, {
      mappers: {
        Publisher: "./entities#PublisherEntity",
        Identifiable: "./entities#IdentifiableEntity",
      },
      scalars: {},
      enumValues: {},
    });
    expect(code).toMatchSnapshot();
  });
});
