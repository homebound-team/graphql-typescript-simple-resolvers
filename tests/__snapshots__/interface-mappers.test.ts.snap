// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Interface Mapped Types snapshots interface mapper generation 1`] = `
"import { GraphQLResolveInfo } from "graphql";
import { Context } from "./context";
import { IdentifiableEntity, PublisherEntity } from "./entities";

export interface Resolvers {
  Query: QueryResolvers;
  SmallPublisher?: SmallPublisherResolvers;
  LargePublisher?: LargePublisherResolvers;
  Book?: BookResolvers;
  Magazine?: MagazineResolvers;
}

export interface PublisherResolvers<T> {
  name: Resolver<T, {}, string | null | undefined>;
  country: Resolver<T, {}, string | null | undefined>;
}

export interface IdentifiableResolvers<T> {
  id: Resolver<T, {}, string>;
}

export type PublisherTypes = SmallPublisher | LargePublisher;

export type IdentifiableTypes = LargePublisher;

export type UnionResolvers = {
  Publisher: { __resolveType(o: SmallPublisher | LargePublisher): string };
  Identifiable: { __resolveType(o: LargePublisher): string };
};

export interface QueryResolvers {
  publishers: Resolver<{}, {}, ReadonlyArray<PublisherEntity>>;
  identifiableItems: Resolver<{}, {}, ReadonlyArray<IdentifiableEntity>>;
  getPublisher: Resolver<{}, QueryGetPublisherArgs, PublisherEntity | null | undefined>;
  books: Resolver<{}, {}, readonly Book[]>;
}

export interface SmallPublisherResolvers extends PublisherResolvers<SmallPublisher> {
  employees: Resolver<SmallPublisher, {}, number | null | undefined>;
}

export interface LargePublisherResolvers
  extends PublisherResolvers<LargePublisher>, IdentifiableResolvers<LargePublisher> {
  revenue: Resolver<LargePublisher, {}, number | null | undefined>;
  subsidiaries: Resolver<LargePublisher, {}, readonly string[]>;
}

export interface BookResolvers {
  title: Resolver<Book, {}, string>;
  publisher: Resolver<Book, {}, PublisherEntity | null | undefined>;
  isbn: Resolver<Book, {}, string>;
}

export interface MagazineResolvers {
  title: Resolver<Magazine, {}, string>;
  publisher: Resolver<Magazine, {}, PublisherEntity>;
  issueNumber: Resolver<Magazine, {}, number>;
}

type MaybePromise<T> = T | Promise<T>;
export type Resolver<R, A, T> = (root: R, args: A, ctx: Context, info: GraphQLResolveInfo) => MaybePromise<T>;

export type SubscriptionResolverFilter<R, A, T> = (
  root: R | undefined,
  args: A,
  ctx: Context,
  info: GraphQLResolveInfo,
) => boolean | Promise<boolean>;
export type SubscriptionResolver<R, A, T> = {
  subscribe: (root: R | undefined, args: A, ctx: Context, info: GraphQLResolveInfo) => AsyncIterator<T>;
};

export interface QueryGetPublisherArgs {
  id: string;
}
export interface SmallPublisher {
  name: string | null | undefined;
  country: string | null | undefined;
  employees: number | null | undefined;
}

export interface LargePublisher {
  id: string;
  name: string | null | undefined;
  country: string | null | undefined;
  revenue: number | null | undefined;
  subsidiaries: string[];
}

export interface Book {
  title: string;
  publisher: PublisherEntity | null | undefined;
  isbn: string;
}

export interface Magazine {
  title: string;
  publisher: PublisherEntity;
  issueNumber: number;
}
"
`;
